awscrt/auth.py

```diff
@@ -196,11 +201,14 @@ def new_profile(
         Returns:
             AwsCredentialsProvider:
         """
-        assert isinstance(client_bootstrap, ClientBootstrap)
+        assert isinstance(client_bootstrap, ClientBootstrap) or client_bootstrap is None
         assert isinstance(profile_name, str) or profile_name is None
         assert isinstance(config_filepath, str) or config_filepath is None
         assert isinstance(credentials_filepath, str) or credentials_filepath is None
 
+        if client_bootstrap is None:
+            client_bootstrap = ClientBootstrap.get_or_create_static_default()
+
```

Comment:
Where would we release the static client bootstrap once it get created here? Currently I only find the release happens in the MQTT Connection, but I think we can also use AwsCredentialsProvider for other purpose.

Reply:
I *think*, and it is entirely possible I am wrong here, that Python's garbage collector would free the singleton-like object once the module is unloaded, so it shouldn't need to be freed manually by the user.

That said though, I'm not entirely sure. I read a bunch about various singleton patterns in Python, but unfortunately the part about how memory is managed was rather unclear.

I probably should take another look at the design and confirm that it's being freed by unloading the module to ensure that it is indeed being garbage collected.

Reply:
I just checked and can confirm that when the module is unloaded and automatically garbage collected, it will free the singleton automatically. I added the following code to the end of the PubSub sample:

~~~
print ("Client bootstrap:")
print (io._static_client_bootstrap)

from importlib import reload
io = reload(io)
print ("Client bootstrap after reload:")
print (io._static_client_bootstrap)
~~~

And it reports that the object is cleared:

~~~
Client bootstrap:
<awscrt.io.ClientBootstrap object at 0x10a86c300>
Client bootstrap after reload:
None
~~~

Reply:
Yeah, Python is a garbage-collected language so you usually don't release or delete anything explicitly, just let the reference drop.

Reply:
also, mild warning about `importlib.reload`. I wouldn't get into the habit of using that with aws-crt-python. The [docs](https://docs.python.org/3/library/importlib.html#importlib.reload) warn:

>  In many cases extension modules are not designed to be initialized more than once, and may fail in arbitrary ways when reloaded.

We haven't really tested reloading the C parts, and if something isn't tested it's probably broken. Years ago I wondered if reloading was something we had to support, but saw that that statement in the docs as a green light to ignore it